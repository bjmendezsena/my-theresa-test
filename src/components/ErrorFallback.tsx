export const ErrorFallback = () => {
  return <div>Ooops, something went wrong :(</div>;
};
