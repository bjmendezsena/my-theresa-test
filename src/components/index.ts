export * from './BackButton';
export * from './CategoryCarousel';
export * from './ErrorFallback';
export * from './ExampleComponent';
export * from './FavouriteMoviesGrid';
export * from './GenreList';
export * from './Hero';
export * from './Layouts';
export * from './MovieCard';
export * from './MovieInfo';
export * from './MovieMeta';
export * from './MovieOverview';
export * from './MoviePoster';
export * from './WishlistButton';
