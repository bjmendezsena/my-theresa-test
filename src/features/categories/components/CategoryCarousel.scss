.category-carousel {
  margin-bottom: $spacing-12;

  &__title {
    font-size: $text-2xl;
    font-weight: $font-bold;
    color: $white;
    margin: 0 0 $spacing-6;
    padding: 0 $spacing-6;

    @include mobile-only {
      font-size: $text-xl;
      padding: 0 $spacing-4;
      margin-bottom: $spacing-4;
    }
  }

  &__container {
    position: relative;
  }

  &__items {
    display: flex;
    gap: $spacing-3;
    overflow-x: auto;
    scroll-behavior: smooth;
    padding: 0 $spacing-6;
    scrollbar-width: none;
    -ms-overflow-style: none;

    &::-webkit-scrollbar {
      display: none;
    }

    @include mobile-only {
      padding: 0 $spacing-4;
      gap: $spacing-2;
    }

    @include tablet-up {
      padding: 0 $spacing-8;
      gap: $spacing-3;
    }

    @include desktop-up {
      padding: 0 $spacing-12;
      gap: $spacing-4;
    }
  }

  &__item {
    flex-shrink: 0;
    max-width: 250px;
    min-width: 180px;

    @include mobile-only {
      min-width: 140px;
      max-width: 160px;
    }

    @include tablet-up {
      min-width: 200px;
      max-width: 240px;
    }

    @include desktop-up {
      min-width: 220px;
      max-width: 280px;
    }

    @include large-desktop-up {
      max-width: 320px;
    }
  }

  &__empty {
    color: $gray-400;
    font-size: $text-base;
    text-align: center;
    padding: $spacing-8 $spacing-4;
    font-style: italic;
  }

  &__nav-btn {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    width: 50px;
    height: 50px;
    border: none;
    background: linear-gradient(
      135deg,
      rgba($black, 0.8),
      rgba($gray-900, 0.9)
    );
    color: $white;
    border-radius: $radius-full;
    cursor: pointer;
    transition: all $duration-300 $ease-out;
    backdrop-filter: blur(10px);
    box-shadow: $shadow-lg;
    z-index: $z-dropdown;
    display: flex;
    align-items: center;
    justify-content: center;

    &:hover:not(&--disabled) {
      background: linear-gradient(
        135deg,
        rgba($primary-600, 0.9),
        rgba($primary-700, 0.9)
      );
      transform: translateY(-50%) scale(1.1);
      box-shadow: $shadow-xl;
    }

    &:active:not(&--disabled) {
      transform: translateY(-50%) scale(0.95);
    }

    &--left {
      left: $spacing-2;

      @include mobile-only {
        left: $spacing-1;
      }
    }

    &--right {
      right: $spacing-2;

      @include mobile-only {
        right: $spacing-1;
      }
    }

    &--disabled {
      opacity: 0;
      pointer-events: none;
      transform: translateY(-50%) scale(0.8);
    }

    @include mobile-only {
      width: 40px;
      height: 40px;
    }
  }

  &__nav-icon {
    font-size: $text-2xl;
    font-weight: $font-bold;
    line-height: 1;
    user-select: none;

    @include mobile-only {
      font-size: $text-xl;
    }
  }

  &__container:hover {
    .category-carousel__nav-btn:not(.category-carousel__nav-btn--disabled) {
      opacity: 1;
    }
  }

  @include tablet-up {
    &__title {
      padding: 0 $spacing-8;
    }

    &__nav-btn {
      &--left {
        left: $spacing-3;
      }

      &--right {
        right: $spacing-3;
      }
    }
  }

  @include desktop-up {
    &__title {
      padding: 0 $spacing-12;
    }

    &__nav-btn {
      &--left {
        left: $spacing-4;
      }

      &--right {
        right: $spacing-4;
      }
    }
  }
}

@media (prefers-reduced-motion: no-preference) {
  .category-carousel__items {
    scroll-behavior: smooth;
  }
}

.category-carousel__nav-btn:focus {
  outline: 2px solid $primary-400;
  outline-offset: 2px;
}

.category-carousel--loading {
  .category-carousel__items {
    display: flex;
    gap: $spacing-3;
  }

  .category-carousel__item {
    width: calc((100vw - 2 * $spacing-6 - 4 * $spacing-4) / 3);
    height: 350px;
    background: linear-gradient(
      90deg,
      $gray-800 25%,
      $gray-700 50%,
      $gray-800 75%
    );
    background-size: 200% 100%;
    animation: shimmer 1.5s infinite;
    border-radius: $radius-lg;
    flex-shrink: 0;

    @include mobile-only {
      width: calc((100vw - 2 * $spacing-4 - 3 * $spacing-3) / 2.5);
      height: 280px;
    }

    @include tablet-up {
      width: calc((100vw - 2 * $spacing-8 - 4 * $spacing-4) / 3);
      height: 320px;
    }

    @include desktop-up {
      width: calc((100vw - 2 * $spacing-12 - 4 * $spacing-4) / 3);
      height: 380px;
    }

    @include large-desktop-up {
      width: calc((1200px - 4 * $spacing-4) / 3);
      height: 400px;
    }
  }
}

@keyframes shimmer {
  0% {
    background-position: 200% 0;
  }
  100% {
    background-position: -200% 0;
  }
}
