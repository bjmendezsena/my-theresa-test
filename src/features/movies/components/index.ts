export * from './BackButton';
export * from './GenreList';
export * from './Hero';
export * from './MovieInfo';
export * from './MovieMeta';
export * from './MovieOverview';
export * from './MoviePoster';
export * from './WishlistButton';
export * from './MovieCard';
