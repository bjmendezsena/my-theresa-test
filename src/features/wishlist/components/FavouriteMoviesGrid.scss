@use '../../../styles/variables' as *;

.favourite-movies-grid {
  padding: $spacing-6;
  max-width: 1200px;
  margin: 0 auto;

  @include mobile-only {
    padding: $spacing-4;
  }

  // Header Section
  &__header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: $spacing-8;
    padding-bottom: $spacing-4;
    border-bottom: 1px solid $gray-800;

    @include mobile-only {
      flex-direction: column;
      align-items: flex-start;
      gap: $spacing-2;
      margin-bottom: $spacing-6;
    }
  }

  &__title {
    font-size: $text-3xl;
    font-weight: $font-bold;
    color: $white;
    margin: 0;
    display: flex;
    align-items: center;
    gap: $spacing-3;

    &::before {
      content: '❤️';
      font-size: $text-2xl;
    }

    @include mobile-only {
      font-size: $text-2xl;
    }
  }

  &__count {
    background-color: rgba($primary-500, 0.2);
    color: $primary-400;
    padding: $spacing-2 $spacing-3;
    border-radius: $radius-full;
    font-size: $text-sm;
    font-weight: $font-semibold;
    border: 1px solid rgba($primary-500, 0.3);

    @include mobile-only {
      align-self: flex-end;
    }
  }

  // Grid Section
  &__grid {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    gap: $spacing-6;

    @include mobile-only {
      grid-template-columns: repeat(auto-fill, minmax(180px, 1fr));
      gap: $spacing-4;
    }

    @include tablet-up {
      grid-template-columns: repeat(auto-fill, minmax(220px, 1fr));
      gap: $spacing-5;
    }

    @include desktop-up {
      grid-template-columns: repeat(auto-fill, minmax(280px, 1fr));
      gap: $spacing-6;
    }

    @include large-desktop-up {
      grid-template-columns: repeat(4, 1fr); // Fixed 4 columns on large screens
    }
  }

  &__item {
    transition: transform $duration-300 $ease-out;

    &:hover {
      transform: translateY(-8px) scale(1.03);
    }
  }

  // Empty State
  &__empty {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    text-align: center;
    padding: $spacing-16 $spacing-4;
    min-height: 400px;

    @include mobile-only {
      padding: $spacing-12 $spacing-4;
      min-height: 300px;
    }
  }

  &__empty-icon {
    font-size: 80px;
    margin-bottom: $spacing-6;
    opacity: 0.5;
    filter: grayscale(100%);

    @include mobile-only {
      font-size: 60px;
      margin-bottom: $spacing-4;
    }
  }

  &__empty-title {
    font-size: $text-2xl;
    font-weight: $font-bold;
    color: $white;
    margin: 0 0 $spacing-4;

    @include mobile-only {
      font-size: $text-xl;
    }
  }

  &__empty-description {
    font-size: $text-base;
    color: $gray-400;
    margin: 0;
    max-width: 400px;
    line-height: $leading-relaxed;

    @include mobile-only {
      font-size: $text-sm;
    }
  }
}

// Loading state
.favourite-movies-grid--loading {
  .favourite-movies-grid__grid {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    gap: $spacing-6;
  }

  .favourite-movies-grid__item {
    height: 320px;
    background: linear-gradient(
      90deg,
      $gray-800 25%,
      $gray-700 50%,
      $gray-800 75%
    );
    background-size: 200% 100%;
    animation: shimmer 1.5s infinite;
    border-radius: $radius-lg;

    @include mobile-only {
      height: 280px;
    }
  }
}

@keyframes shimmer {
  0% {
    background-position: 200% 0;
  }
  100% {
    background-position: -200% 0;
  }
}

// Responsive grid adjustments
@media (min-width: 1400px) {
  .favourite-movies-grid {
    &__grid {
      grid-template-columns: repeat(5, 1fr); // 5 columns on very large screens
    }
  }
}

// Animation enhancements
.favourite-movies-grid__item {
  animation: fadeInUp 0.6s ease-out backwards;

  @for $i from 1 through 20 {
    &:nth-child(#{$i}) {
      animation-delay: #{$i * 0.1}s;
    }
  }
}

@keyframes fadeInUp {
  from {
    opacity: 0;
    transform: translateY(30px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}
