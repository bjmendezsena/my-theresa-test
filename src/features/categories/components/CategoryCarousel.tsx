import { useState, useRef, useEffect } from 'react';
import { Category } from '@/features/categories';
import { useDiscoverMovies } from '@/features/movies';
import { IMAGE_BASE_URL, MovieCard } from '@/features/movies';
import './CategoryCarousel.scss';

export interface CategoryCarouselProps {
  category: Category;
}

export const CategoryCarousel = ({ category }: CategoryCarouselProps) => {
  const { data } = useDiscoverMovies({
    genres: [category.id],
  });

  const { results = [] } = data || {};
  const carouselRef = useRef<HTMLDivElement>(null);
  const [canScrollLeft, setCanScrollLeft] = useState(false);
  const [canScrollRight, setCanScrollRight] = useState(true);

  const scrollAmount = 300; // Amount to scroll in pixels

  const updateScrollButtons = () => {
    if (!carouselRef.current) return;

    const { scrollLeft, scrollWidth, clientWidth } = carouselRef.current;
    setCanScrollLeft(scrollLeft > 0);
    setCanScrollRight(scrollLeft < scrollWidth - clientWidth - 1);
  };

  const scrollLeft = () => {
    if (!carouselRef.current) return;
    carouselRef.current.scrollBy({
      left: -scrollAmount,
      behavior: 'smooth',
    });
  };

  const scrollRight = () => {
    if (!carouselRef.current) return;
    carouselRef.current.scrollBy({
      left: scrollAmount,
      behavior: 'smooth',
    });
  };

  useEffect(() => {
    const carousel = carouselRef.current;
    if (!carousel) return;

    // Initial check
    updateScrollButtons();

    // Listen for scroll events
    carousel.addEventListener('scroll', updateScrollButtons);

    // Listen for resize events
    window.addEventListener('resize', updateScrollButtons);

    return () => {
      carousel.removeEventListener('scroll', updateScrollButtons);
      window.removeEventListener('resize', updateScrollButtons);
    };
  }, [results]);

  if (results.length === 0) {
    return (
      <div className="category-carousel">
        <h2 className="category-carousel__title">{category.name}</h2>
        <p className="category-carousel__empty">
          No movies found in this category.
        </p>
      </div>
    );
  }

  return (
    <div className="category-carousel">
      <h2 className="category-carousel__title">{category.name}</h2>

      <div className="category-carousel__container">
        <button
          className={`category-carousel__nav-btn category-carousel__nav-btn--left ${
            !canScrollLeft ? 'category-carousel__nav-btn--disabled' : ''
          }`}
          onClick={scrollLeft}
          disabled={!canScrollLeft}
          aria-label="Scroll left"
        >
          <span className="category-carousel__nav-icon">‹</span>
        </button>

        <div
          ref={carouselRef}
          className="category-carousel__items"
          onScroll={updateScrollButtons}
        >
          {results.map(movie => (
            <MovieCard
              key={movie.id}
              id={movie.id}
              title={movie.title}
              posterPath={`${IMAGE_BASE_URL}${movie.poster_path}`}
              releaseDate={movie.release_date?.toString()}
              voteAverage={movie.vote_average}
              overview={movie.overview}
              className="category-carousel__item"
            />
          ))}
        </div>

        <button
          className={`category-carousel__nav-btn category-carousel__nav-btn--right ${
            !canScrollRight ? 'category-carousel__nav-btn--disabled' : ''
          }`}
          onClick={scrollRight}
          disabled={!canScrollRight}
          aria-label="Scroll right"
        >
          <span className="category-carousel__nav-icon">›</span>
        </button>
      </div>
    </div>
  );
};
