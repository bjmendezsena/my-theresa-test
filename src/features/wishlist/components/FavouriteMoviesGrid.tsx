import { MovieCard } from '@/features/movies';
import { useWishlist } from '@/features/wishlist';
import './FavouriteMoviesGrid.scss';

export const FavouriteMoviesGrid = () => {
  const { items } = useWishlist();

  if (items.length === 0) {
    return (
      <div className="favourite-movies-grid">
        <div className="favourite-movies-grid__empty">
          <div className="favourite-movies-grid__empty-icon">❤️</div>
          <h3 className="favourite-movies-grid__empty-title">
            No movies in your wishlist yet
          </h3>
          <p className="favourite-movies-grid__empty-description">
            Start adding movies to your wishlist to see them here!
          </p>
        </div>
      </div>
    );
  }

  return (
    <div className="favourite-movies-grid">
      <div className="favourite-movies-grid__header">
        <h2 className="favourite-movies-grid__title">My Wishlist</h2>
        <span className="favourite-movies-grid__count">
          {items.length} {items.length === 1 ? 'movie' : 'movies'}
        </span>
      </div>

      <div className="favourite-movies-grid__grid">
        {items.map(movie => (
          <MovieCard
            key={movie.id}
            id={movie.id}
            title={movie.title}
            posterPath={movie.posterPath}
            releaseDate={movie.releaseDate?.toString()}
            voteAverage={movie.rating}
            overview={movie.overview}
            className="favourite-movies-grid__item"
            showWishlistButton
          />
        ))}
      </div>
    </div>
  );
};
