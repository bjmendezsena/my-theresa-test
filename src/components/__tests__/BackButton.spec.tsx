import {
  renderWithProviders,
  expect,
  describe,
  test,
  act,
} from '@/tests/utils';
import { BackButton } from '@/components';

describe(`<${BackButton.name}/>`, () => {
  const getComponent = () =>
    renderWithProviders(<BackButton to="/">Go back</BackButton>);

  test('renders correctly', () => {
    const { container } = getComponent();
    expect(container).toMatchSnapshot();
  });

  test('calls history.back on click', () => {
    const { getByRole } = getComponent();
    const backButton = getByRole('button');

    act(() => {
      backButton.click();
    });

    expect(window.history.length).toBe(1);
  });
});
