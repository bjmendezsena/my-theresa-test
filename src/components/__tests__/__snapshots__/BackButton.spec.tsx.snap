// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<BackButton/> > renders correctly 1`] = `
<div>
  <a
    class="back-button "
    data-discover="true"
    href="/"
    role="button"
  >
    Go back
  </a>
</div>
`;
